Begin overhauling libtock_platform's syscalls trait for Tock 2.0.

jrvanwhy opened the pull request:
I made the following changes to the design of `Syscalls`:

1. The trait is now designed for the (work-in-progress) Tock 2.0 syscalls rather
   than the Tock 1.0 syscalls.
2. The higher-level API that I previously intended to implement in the Platform
   type is now expressed in the form of provided functions in Syscalls.

So far, I have only implemented `yield`. I'm sending this for review now because it is already getting large, and I'd prefer to split it up into several PRs rather than send a single huge PR.

There is still significant design work to be done in terms of error handling for `subscribe`, `command`, both `allow`s, and `memop`.

EDIT: To everyone reviewing this, here is a [direct link to syscalls.rs](https://github.com/jrvanwhy/libtock-rs/blob/syscalls-2.0/core/platform/src/syscalls.rs), in case you want to read it *without* seeing the diff from the previous version.

jrvanwhy commented:
@lschuermann and @phil-levis, I added you as reviewers because -- to an extent -- this trait represents my understanding of the Tock 2.0 TRD.

jrvanwhy commented:
I sent this for review, then came up with a possibly-better idea for handling the two API layers problem. Because today and tomorrow are a major US holiday, I'll probably try out my idea on Monday. Here's the idea:

I can move the "low-level" API (`raw_yield`, `four_arg_syscall`, `zero_arg_memop`) into a `RawSyscalls` trait that is implemented by `libtock_runtime::TockSyscalls` and `libtock_unittest::FakeSyscalls`. The high-level API remains in the `Syscalls` trait. I have an impl like the following in `libtock_platform` to implement `Syscalls`:

```rust
impl<T: RawSyscalls> Syscalls for T {
    // Omitted
}
```

That way, drivers can still depend on `Syscalls` but I remove the confusion of having two APIs. It also cuts down on the file size of `syscalls.rs`.

jrvanwhy commented:
I got bored and decided to implement the idea from my previous comment. I have now split `RawSyscalls` into `Syscalls` and `RawSyscalls`. Apart from moving things around, the only substantive technical change is that components that depend on a `Syscalls` implementation no longer have access to the raw system calls. Code that needs access to raw system calls (which I do not anticipate at the moment) would need to depend on `RawSyscalls` instead of `Syscalls`.

hudson-ayers reviewed (approved):
I'm no expert on the Tock 2.0 system calls, but this looks like a nice design to me.

I noticed this uses only ARM register names in the comments describing the behavior of each system call -- perhaps somewhere should note r0 refers to r0 on ARM, but a0 on risc-v, etc.?

jrvanwhy commented:
> I noticed this uses only ARM register names in the comments describing the behavior of each system call -- perhaps somewhere should note r0 refers to r0 on ARM, but a0 on risc-v, etc.?

I actually copied that from the 2.0 syscalls TRD, which uses the ARM register names. I added a comment near the top of the file pointing that out.

hudson-ayers commented:
> I actually copied that from the 2.0 syscalls TRD, which uses the ARM register names. I added a comment near the top of the file pointing that out.

Your latest commit doesn't seem to actually touch `raw_syscalls.rs`, so you may have forgotten to add a changed file

jrvanwhy commented:
> > I actually copied that from the 2.0 syscalls TRD, which uses the ARM register names. I added a comment near the top of the file pointing that out.
> 
> Your latest commit doesn't seem to actually touch `raw_syscalls.rs`, so you may have forgotten to add a changed file

Good catch! I recall typo-ing a `git` command while making that change, I bet I blew away my `raw_syscalls.rs` changes without noticing. Fixed with another commit.

lschuermann commented on core/platform/src/raw_syscalls.rs:
```diff
@@ -0,0 +1,194 @@
+// Memop takes 1 or 2 arguments (operation and an optional argument). Because it
+// is part of the core kernel, it is okay for us to leave arbitrary data in the
+// argument register for operations where the argument register is unused
+// (again, in line with Tock's threat model). Memop returns up to 2 return
```
I guess by "returns [...] arguments" you are referring to the system call return value? As per the [current TRD state](https://github.com/tock/tock/blob/5e2e7db4d831661e9f4066cec18ac838d95882e8/doc/reference/trd-syscalls.md#46-memop-class-id-5), the "[...] success return type is Memop class system call specific", indicating that depending on the specifc Memop call, we might use the full 4 registers used for the various success return type variants.

I can imagine that this will not only be allowed theoretically, but that we might update Memop to take advantage of the gained registers and pass values commonly used together (such as the app's flash start and end address) in a single system call. We're going to touch & port Memop very soon, now that the required infrastructure is in place.

jrvanwhy replied:
I dug through all [existing memop types](https://github.com/tock/tock/blob/master/doc/syscalls/memop.md) and determined that none of them use any success types other than Success and Success with u32. My statement that "Memop returns up to 2 return arguments" is based on the existing memop types, and was not meant to be a theoretical limitation on all memop operations. I expect to redesign this interface (by adding r2 and/or r3 as return registers) if that changes.

I will update this comment to make that more clear.

lschuermann replied:
Great, thanks. I think it will be a good idea to document the return variants of the individual Memop operations once the design stabilizes, as part of the regular syscalls documentation in the kernel `docs/`. It's good to know that we are on the same page here and keeping this in mind.

jrvanwhy replied:
I just pushed commit [72a967](https://github.com/tock/libtock-rs/pull/261/commits/72a967363995cd0e805edfb35876aad9063a6d36), which updates the comments around memop. It should be more clear now that adding support for future memop operations may require changing clobbers.

I'm not too concerned about someone missing the clobber changes because adding a new return type would require changing the return type for `zero_arg_memop` and/or `one_arg_memop`. That return type change will force changes to the `RawSyscalls` implementation, and the obvious way to implement those changes is to implement correct output register specifications in the assembly.

lschuermann commented on core/platform/src/raw_syscalls.rs:
```diff
@@ -0,0 +1,194 @@
+    //        registers. Returns r0-r3 in order.
+    //     3. Does not mark any registers as clobbered.
+    //     4. Has all of the following options:
+    //            preserves_flags  (these system calls do not touch flags)
```
I'm guessing this is correct, in that on ARM it's only referring to the ARM Current Program Status Register, which is saved to the respective Saved Processor Status Register of the user mode and then saved in kernel and restored accordingly (though I'm not as experienced with ARM Cortex-M). Is there something else to watch out for here?

[The Rust inline-asm RFC](https://github.com/rust-lang/rfcs/blob/master/text/2873-inline-asm.md#rules-for-inline-assembly) doesn't indicate other registers in need of restoring to me (except if an userspace app is compiled for hard float), but I'd rather have an expert(TM) confirm. :smiley:

jrvanwhy replied:
By my reading, on 32-bit ARM you need to preserve `CPSR` and `FPSCR` (if hardware float available), and on RISC-V you need to preserve `fcsr` (if hardware float is available).

lschuermann commented on core/platform/src/raw_syscalls.rs:
```diff
@@ -0,0 +1,194 @@
+
+    // zero_arg_memop is used to invoke memop operations that do not accept an
+    // argument register. Because there are no memop commands that set r2 or r3,
+    // this only needs to return r0 and r1.
```
Same point as above.

lschuermann commented on core/platform/src/raw_syscalls.rs:
```diff
@@ -0,0 +1,194 @@
+
+    // one_arg_memop is used to invoke memop operations that take an argument.
+    // Because there are no memop operations that set r2 or r3, this only needs
+    // to return r0 and r1.
```
Same point as above.

lschuermann commented on core/platform/src/raw_syscalls.rs:
```diff
@@ -0,0 +1,194 @@
+
+#[non_exhaustive]
+#[repr(usize)]
+pub enum OneArgMemop {
```
It's really cool to guarantee that each `memop` operand can only be passed into the respective wrapper.

jrvanwhy replied:
This also allows the kernel to add new `memop` types that violate the assumptions I made in this file without making `RawSyscalls` an unsound API.

lschuermann commented:
This is great! I'm by no means an expert on `libtock-rs`, but it seems to capture the essence of the system call interface well. I like the design in two stages, with low-level assembly wrappers and higher-level Rust wrapper functions. I think we can take a lot of inspiration from this (especially the asm clobbers) for `libtock-c` as well.

jrvanwhy commented:
I just pushed a commit that replaces several uses of `usize` in the trait with `u32`.

I initially used `usize` for all register values because `usize` is register-sized on all platforms, but I realized that isn't all that relevant to `RawSyscalls`' design. Tock's syscall ABI is only specified for 32-bit systems, so `u32` is a perfectly valid type for representing a register as well. For system call inputs and outputs that never carry pointer-sized values (e.g. `r0`), using `u32` has the advantage of removing error checking from the unit test `RawSyscalls` implementation.

I changed my documentation about the use of `usize` to represent registers, and replaced it with a simple rule: register values that can contain pointers are `usize`s and values that cannot are `u32`'s. I updated the types throughout `RawSyscalls` to match the new design.

I realize this is a change late in the review cycle of this PR. Feel free to let me know if you think it's a regression and I can back it out.

phil-levis reviewed (commented):
Overall, I think this design seems sound. My one question is whether a slightly different refactoring of raw_syscall.rs might make more sense. Right now, there's specialized calls for yield and two different memops, then a general 4-argument. Would it make sense to have general 0, 1, 2, 3, and 4 arguments, then specialize them to certain calls? E.g., have a "command" which is just a wrapper around a 4-arg?

jrvanwhy commented:
> Overall, I think this design seems sound. My one question is whether a slightly different refactoring of raw_syscall.rs might make more sense. Right now, there's specialized calls for yield and two different memops, then a general 4-argument. Would it make sense to have general 0, 1, 2, 3, and 4 arguments, then specialize them to certain calls?

Here's the thought process I used to arrive at this design:

1. Yield must be its own method, because its clobbers are far more extensive than the other calls.
1. Everything else could be done with generic 0-arg, 1-arg, 2-arg, 3-arg, and 4-arg calls. At this point, I have 6 system calls in my design.
1. The 0-arg and 3-arg calls are never used, so I removed them from the list. Four system calls remain.
1. The 1-arg and 2-arg calls are only used for `memop`. By limiting them to `memop`, I am able to give them a safe API (i.e. they are not `unsafe` to call). So I renamed them to `zero_arg_memop` and `one_arg_memop`. 4-arg is the only remaining `unsafe` method at this point.

This process left me with 4 methods: yield, 4-arg, `zero_arg_memop`, and `one_arg_memop`.

If Tock adds a new system call type that has 1 or 2 arguments, I would replace the corresponding `_arg_memop` method with a more generic -- but potentially `unsafe` -- system call method.

> E.g., have a "command" which is just a wrapper around a 4-arg?

I'm not quite sure what you mean about this part. `Syscalls` will have a `command` which is a wrapper around `RawSyscalls::four_arg_syscall`, as well as `subscribe` and allow wrappers.

phil-levis commented:
This is good reasoning -- explaining it a bit (so people understand why the design is what it is) would be good, perhaps in a TRD? Often the *why* is just as important as the *what*. It helps make sure that people don't miss something you figured out and make mistakes you avoided. :)

phil-levis commented:
> > E.g., have a "command" which is just a wrapper around a 4-arg?
> 
> I'm not quite sure what you mean about this part. `Syscalls` will have a `command` which is a wrapper around `RawSyscalls::four_arg_syscall`, as well as `subscribe` and allow wrapper

The thought was to hide the `u8`, in the way it's hidden for yield and memop. But it's not a big deal.

jrvanwhy commented:
> This is good reasoning -- explaining it a bit (so people understand why the design is what it is) would be good, perhaps in a TRD? Often the _why_ is just as important as the _what_. It helps make sure that people don't miss something you figured out and make mistakes you avoided. :)

The huge comment above `RawSyscalls` was intended to explain the _why_ behind its design. I just reworked it to look more like the explanation I just gave (see the new version [here](https://github.com/jrvanwhy/libtock-rs/blob/syscalls-2.0/core/platform/src/raw_syscalls.rs#L10)).

jrvanwhy commented:
Phil, if you are happy with my revised comment, then I think this PR is ready to merge.

If you believe the comment should be pulled out into a Markdown file (like a TRD) then I'm happy to do so, but I would prefer to do that in a separate PR so that I can start sending out PRs that depend on `RawSyscalls` now.

phil-levis commented:
> Phil, if you are happy with my revised comment, then I think this PR is ready to merge.
> 
> If you believe the comment should be pulled out into a Markdown file (like a TRD) then I'm happy to do so, but I would prefer to do that in a separate PR so that I can start sending out PRs that depend on `RawSyscalls` now.

I'm generally a fan of detailed design documentation living separately from code, but I'm not going to demand it. I think this is fine to merge now.

lschuermann commented:
Regarding the `usize` to `u32` conversion: You are right in the fact that the current ABI is only specified for 32-bit architectures. However we did think about how an ABI could work on 64-bit architectures (which, again, we should not worry too much about right now):

In essence, Command's return values would still be limited to a width of 32 bits. This is to avoid leaking these architectural differences into capsule code, which then would need to behave differently with regards to the interface to processes (ABI). The return values of Subscribe and Allow, having the fields _address_, _length_ (in the case of Allow), and _userdata_ (in the case of Subscribe), would be 64-bit sized. While _address_ and _length_ must be of native integer width, the reasoning behind _userdata_ having native integer width is that applications may use it to pass pointers, and this information is hidden in a high-level Rust types, such that capsules cannot access it.

Following this reasoning, the current design seems to encapsulate the ideas well and is likely to be compatible with a future 64-bit ABI, should it be defined.

hudson-ayers commented:
bors r+
